pkg/registry/core/service/storage/storage_test.go

```diff
@@ -561,29 +561,63 @@ func TestServiceDefaultOnRead(t *testing.T) {
-		expect: svctest.MakeService("foo", svctest.SetClusterIPs("10.0.0.1")),
+		name:  "no change v6",
+		input: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1")),
+		expect: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+			svctest.SetIPFamilies(api.IPv6Protocol)),
 	}, {
-		name:   "no change v6",
-		input:  svctest.MakeService("foo", svctest.SetClusterIPs("2000::1")),
-		expect: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1")),
+		name:  "missing clusterIPs v6",
+		input: svctest.MakeService("foo", svctest.SetClusterIP("2000::1")),
+		expect: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+			svctest.SetIPFamilies(api.IPv6Protocol)),
 	}, {
-		name:   "missing clusterIPs v6",
-		input:  svctest.MakeService("foo", svctest.SetClusterIP("2000::1")),
-		expect: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1")),
+		name:  "list, no change v4",
+		input: makeServiceList(svctest.SetClusterIPs("10.0.0.1")),
+		expect: makeServiceList(svctest.SetClusterIPs("10.0.0.1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+			svctest.SetIPFamilies(api.IPv4Protocol)),
 	}, {
-		name:   "list, no change v4",
-		input:  makeServiceList(svctest.SetClusterIPs("10.0.0.1")),
-		expect: makeServiceList(svctest.SetClusterIPs("10.0.0.1")),
+		name:  "list, missing clusterIPs v4",
+		input: makeServiceList(svctest.SetClusterIP("10.0.0.1")),
+		expect: makeServiceList(svctest.SetClusterIPs("10.0.0.1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+			svctest.SetIPFamilies(api.IPv4Protocol)),
 	}, {
-		name:   "list, missing clusterIPs v4",
-		input:  makeServiceList(svctest.SetClusterIP("10.0.0.1")),
-		expect: makeServiceList(svctest.SetClusterIPs("10.0.0.1")),
+		name:   "external name",
+		input:  makeServiceList(svctest.SetTypeExternalName),
+		expect: makeServiceList(svctest.SetTypeExternalName),
+	}, {
+		name:  "dual, v4v6",
+		input: svctest.MakeService("foo", svctest.SetClusterIPs("10.0.0.1", "2000::1")),
+		expect: svctest.MakeService("foo", svctest.SetClusterIPs("10.0.0.1", "2000::1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicyRequireDualStack),
+			svctest.SetIPFamilies(api.IPv4Protocol, api.IPv6Protocol)),
+	}, {
+		name:  "dual, v6v4",
+		input: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1", "10.0.0.1")),
+		expect: svctest.MakeService("foo", svctest.SetClusterIPs("2000::1", "10.0.0.1"),
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicyRequireDualStack),
+			svctest.SetIPFamilies(api.IPv6Protocol, api.IPv4Protocol)),
+	}, {
+		name:  "headless",
+		input: svctest.MakeService("foo", svctest.SetHeadless),
+		expect: svctest.MakeService("foo", svctest.SetHeadless,
+			svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+			svctest.SetIPFamilies(api.IPv4Protocol)),
 	}, {
```

Comment:
diff --git a/pkg/registry/core/service/storage/storage_test.go b/pkg/registry/core/service/storage/storage_test.go
index 3bbeb7df285..e16c4f82103 100644
--- a/pkg/registry/core/service/storage/storage_test.go
+++ b/pkg/registry/core/service/storage/storage_test.go
@@ -619,9 +619,28 @@ func TestServiceDefaultOnRead(t *testing.T) {
                        svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
                        svctest.SetIPFamilies(api.IPv4Protocol)),
        }, {
-               name:  "not Service or ServiceList",
-               input: &api.Pod{},
-       }}
+               name: "headless-selector-ipfamilypolicy-requiredual",
+               input: svctest.MakeService("foo", svctest.SetHeadless,
+                       svctest.SetIPFamilyPolicy(api.IPFamilyPolicyRequireDualStack),
+               ),
+               expect: svctest.MakeService("foo", svctest.SetHeadless,
+                       svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+                       svctest.SetIPFamilies(api.IPv4Protocol)),
+       }, {
+               name: "headless-selectorless-ipfamilypolicy-single",
+               input: svctest.MakeService("foo", svctest.SetHeadless,
+                       svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+                       svctest.SetSelector(map[string]string{}),
+               ),
+               expect: svctest.MakeService("foo", svctest.SetHeadless,
+                       svctest.SetIPFamilyPolicy(api.IPFamilyPolicySingleStack),
+                       svctest.SetIPFamilies(api.IPv4Protocol)),
+       },
+               {
+
+                       name:  "not Service or ServiceList",
+                       input: &api.Pod{},
+               }}

Reply:
The first case SHOULD be impossible - how would an  object have been stored that way but with no ipFamilies?  I am not  seeing it.

Reply:
oh, I see , so 
```
if len(service.Spec.IPFamilies) > 0 {
return
}
```
means that the IPFamilyPolicy has to be nil after that

Reply:
Yes.  Will add comments to clarify more. :)